consensus/hotstuff/vote_collector.go

```diff
@@ -43,11 +54,19 @@ type VoteCollector interface {
 	Status() VoteCollectorStatus
 }
 
-// VerifyingVoteCollector must implement the same interface as BlockSigner, such that when eventhandler
-// asks voter to vote for the block, the voter will ask VoteAggregator(via BlockSigner interface)
-// to sign the block, and VoteAggregator will read the VerifyingVoteCollector from the vote collectors
-// map and produce the vote.
-// Note CachingVoteCollector can't create vote, only VerifyingVoteCollector can
-type VerifyingVoteCollector interface {
-	BlockSigner
+type VoteCollector interface {
+	VoteCollectorState
+	// ChangeProcessingStatus changes the VoteCollector's internal processing
+	// status. The operation is implemented as an atomic compare-and-swap, i.e. the
```

Comment:
It's not clear to me how the compare and swap works.

The VoteCollectors maintains a map from block ID to collectors. The collectors for different blocks could be in different status. 

When VoteCollectors processes a block, it's the moment that a collector's state could possibly change from one to another. If we change the status of a collector from caching to verifying, we will swap the instance in the map. Meaning, we will create a new verifying collector instance to replace the caching collector instance stored in the map. And VoteCollectors will ensure the concurrent access to the map. Therefore the atomicity of this compare-and-swap must be implemented by the `VoteCollectors`, rather than the VoteCollector. 

So I wonder if the ChangeProcessingStatus is just an internal function of `VoteCollectors`?

Reply:
not exactly, `VoteCollectors` which actually store instances of `VoteCollectorStateMachine` under `hotstuff.VoteCollector` interface, each collector is a state machine that can change it's internal state(Caching, Verifying, etc) in this case all users of `VoteCollectors` operate on same `hotstuff.VoteCollector` which can change state of course since internally it's a state machine. This way we don't replace any instances in map or in users of this class.

Reply:
I see. So basically it's this relation:

```
VoteAggregator
    |--- VoteCollectors
        |-- map[blockID]VoteCollectorStateMachine
                        |----  atomic.Value(VoteCollector in one of the 3 status)
```


OK, makes sense to me.